Create projections client

fchaillou opened the pull request:
Hello,

For work, we needed a projection client so i created one and i'm trying here to upstream it.
It is based on akka http.

This is the exposed API :
```
class ProjectionsClient {
def createProjection(name: String, javascriptCode: String, mode: ProjectionMode = Continuous, allowEmit: Boolean = true): Future[ProjectionCreationResult]

  def fetchProjectionStatus(name: String): Future[ProjectionStatus]

  def fetchProjectionState(name: String, projection: Option[String] = None): Future[Option[JsValue]]

  def fetchProjectionResult(name: String, projection: Option[String] = None): Future[Option[JsValue]]

  def waitForProjectionStatus(name: String, expectedState : ProjectionStatus, waitDuration : FiniteDuration = 10 seconds) : Future[Boolean]

def startProjection(name: String): Future[Unit]

  def stopProjection(name: String): Future[Unit]

  def deleteProjection(name: String): Future[ProjectionDeleteResult]
}
```
Hopes it will fit here !
Fabien

t3hnar commented:
Hi @fchaillou thanks for your effort, will take a look shortly. I'm busy right now with migration to `akka-streams` in order to fix some performance/stability issues.

Btw, I will remove spray dependency and switch clustering to `akka-http` as well.

fchaillou commented:
Hi,
FYI, i just rebased my PR against master.
Don't hesitate if you have any question/remarks !
Thanks

t3hnar commented on src/main/scala/eventstore/EsProjectionsClient.scala:
```diff
@@ -0,0 +1,280 @@
+    val uri = Uri(settings.http.url)
+    val httpExt = Http(system)
+
+    val connectionPool = if (settings.http.protocol == "http") {
```
what if somebody override protocol in `http.url` ?

t3hnar commented on src/main/scala/eventstore/EsProjectionsClient.scala:
```diff
@@ -0,0 +1,280 @@
+}
+
+object EsProjectionsClient {
+  sealed trait ProjectionMode
```
I'd prefer nesting impls inder sealed trait companion object

fchaillou replied:
Makes sense !
Actually, this is kind of a current debate at work on this !
I agree that putting it inside the companion object avoids potential name clash

t3hnar commented on src/main/scala/eventstore/EsProjectionsClient.scala:
```diff
@@ -0,0 +1,280 @@
+
+object EsProjectionsClient {
+  sealed trait ProjectionMode
+  object Continuous extends ProjectionMode
```
case object and for all others

fchaillou replied:
Mistake on my part :)

t3hnar commented on src/main/scala/eventstore/EsProjectionsClient.scala:
```diff
@@ -0,0 +1,280 @@
+    singleRequestWithErrorHandling(request)
+      .mapAsync(1) {
+        case response if response.status == StatusCodes.NotFound => Future.successful(Absent)
+        case response => Unmarshal(response).to[String].map { rawJson =>
```
maybe it does worth to create some case class for Projection with generic fields as status?

fchaillou replied:
I tool all the fields :
```

t3hnar replied:
yep

t3hnar commented on src/main/scala/eventstore/EsProjectionsClient.scala:
```diff
@@ -0,0 +1,280 @@
+   * @return true is the expectedStatus was reached, false otherwise
+   */
+  def waitForProjectionStatus(name: String, expectedStatus: ProjectionStatus, waitDuration: FiniteDuration = 10.seconds): Future[Boolean] = {
+    val retryCount = 10
```
should be a method argument

t3hnar commented on src/main/scala/eventstore/EsProjectionsClient.scala:
```diff
@@ -0,0 +1,280 @@
+    val retryCount = 10
+    val retryDelay = waitDuration / retryCount
+
+    Source(1 to retryCount)
```
overall I find this construction pretty generic, do we really want to have it supported in EventStore ?

fchaillou replied:
I actually only used that for tests so i moved it into my tests.
I thought i could be useful in the API but i agree it can be kept out

t3hnar commented on src/main/scala/eventstore/Settings.scala:
```diff
@@ -84,4 +87,17 @@ object Settings {
\ No newline at end of file
+}
+
+case class HttpSettings(protocol: String = "http", port: Int = 2113, prefix: String = "", url: String = "http://127.0.0.1:2113/") {
```
I think it should look like `case class HttpSettings(uri: Uri = Uri("http://127.0.0.1:2113"))`

fchaillou replied:
Agreed, clearly better this way !

t3hnar commented on src/main/scala/eventstore/EsProjectionsClient.scala:
```diff
@@ -0,0 +1,280 @@
+   * @param allowEmit
+   * @return
+   */
+  def createProjection(name: String, javascriptCode: String, mode: ProjectionMode = Continuous, allowEmit: Boolean = true): Future[ProjectionCreationResult] = {
```
rename `javascriptCode` -> `javascript` ?

fchaillou replied:
Done

t3hnar commented on src/main/scala/eventstore/EsProjectionsClient.scala:
```diff
@@ -0,0 +1,280 @@
+          case ex => throw new ProjectionException(s"Failed to query eventstore on ${request.uri}", ex)
+        }.get
+        if (response.status == StatusCodes.Unauthorized)
+          throw new CannotEstablishConnectionException("Invalid credentials")
```
AccessDeniedException ?

fchaillou replied:
Updated

t3hnar commented on src/main/scala/eventstore/EventStoreExtension.scala:
```diff
@@ -1,15 +1,20 @@
 
   def settings: Settings = Settings(system.settings.config)
 
+  implicit val materializer = ActorMaterializer.create(system)
```
I'd rather not make this as part of public API

fchaillou replied:
changed to be part of the EsProjectionsClient.

fchaillou commented:
Hello,

Just updated my PR, i think i adressed all your comments.
Don't hesitate it you have any other issue.

By the way, do you feel EsProjectionClient being a good name or ProjectionClient to be enough ?
Internally we moved to the latter so i could update my PR aswell if you feel that it make sense.

t3hnar commented:
`ProjectionClient` looks better

fchaillou commented:
Done, actually it is `ProjectionsClient` :)

t3hnar commented:
@fchaillou thx, merged
